File: LeetCode/May challenge/Day 27/Day_27.py
class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        like_dict = {}
        for pair in dislikes:
            if pair[0] in like_dict:
                like_dict[pair[0]].add(pair[1])
            else:
                like_dict[pair[0]] = set([pair[1]])
            if pair[1] in like_dict:
                like_dict[pair[1]].add(pair[0])
            else:
                like_dict[pair[1]] = set([pair[0]])
                
        room_wise = {}
        for i in range(1, N + 1):
            if i not in room_wise:
                room_wise[i] = 0
                stack = [i]
                while stack:
                    a = stack.pop()
                    if a in like_dict:
                        for b in like_dict[a]:
                            if b in room_wise:
                                if room_wise[a] == room_wise[b]:
                                    return False
                            else:
                                room_wise[b] = (room_wise[a] + 1) % 2
                                stack.append(b)
        return True

File: LeetCode/May challenge/Day 27/Day_27.cpp
#define WHITE 0
#define RED 1
#define BLUE 2

class Solution 
{
public:
    bool possibleBipartition(int N, vector<vector<int>> &edges) 
    {
        vector<vector<int>> adj(N + 1); // adjacency list for undirected graph
        vector<int> color(N + 1, WHITE); // color of each vertex in graph, initially WHITE
        vector<bool> explored(N + 1, false); // to check if each vertex has been explored exactly once
        
        // create adjacency list from given edges
        for (auto &edge: edges)
        {
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        
        // queue to perform BFS over each connected component in the graph
        // while performing BFS, we check if we encounter any conflicts while
        // coloring the vertices of the graph
        // conflicts indicate that bi-partition is not possible
        queue<int> q;
        
        for (int i = 1; i <= N; ++i)
        {
            if (!explored[i])
            {
                // this component has not been colored yet
		// we color the first vertex RED and push it into the queue
                color[i] = RED;
                q.push(i);
                
                // perform BFS from vertex i
                while (!q.empty())
                {
                    int u = q.front();
                    q.pop();
                    
                    // check if u is already explored 
                    if (explored[u])
                        continue;
                    explored[u] = true;
                    
                    // for each neighbor of u, execute this loop
                    for (auto v: adj[u])
                    {
                        // v is u's neighboring vertex
                        
                        // checking if there's any conflict in coloring
                        if (color[v] == color[u])
                        {
			// conflict edge found, so we return false 
			// as bi-partition will not be possible
                            return false;
                        }
                        
                        // we color v with the opposite color of u
                        if (color[u] == RED)
                            color[v] = BLUE;
                        else 
                            color[v] = RED;
                        
                        q.push(v);
                    }
                }
            }
        }
        
        // if no conflicts encountered then graph must be bipartite
        // so we return true
        
        return true;
    }
};
